src/Build.UnitTests/Evaluation/Evaluator_Tests.cs

```diff
@@ -4528,6 +4483,67 @@ public void VerifyMSBuildLogsAMessageWhenLocalPropertyCannotOverrideValueOfGloba
                 ResourceUtilities.FormatResourceStringStripCodeAndKeyword("OM_GlobalProperty", "Foo"));
         }
 
+        [Fact]
+        public void VerifyPropertyTrackingLogging()
+        {
+           string testTargets = ObjectModelHelpers.CleanupFileContents(@"
+                                <Project>
+                                     <PropertyGroup>
+                                         <Prop>$(DOES_NOT_EXIST)</Prop>
+                                         <EnvVar>$(DEFINED_ENVIRONMENT_VARIABLE2)</EnvVar>
+                                         <DEFINED_ENVIRONMENT_VARIABLE>Overwritten!</DEFINED_ENVIRONMENT_VARIABLE>
+                                         <NotEnvVarRead>$(DEFINED_ENVIRONMENT_VARIABLE)</NotEnvVarRead>
+                                         <Prop2>Value1</Prop2>
+                                         <Prop2>Value2</Prop2>
+                                     </PropertyGroup>
+                                     <Target Name='Build' />
+                                </Project>");
+
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MsBuildLogPropertyTracking", "1");
+                env.SetEnvironmentVariable("DEFINED_ENVIRONMENT_VARIABLE", "It's Defined!");
+                env.SetEnvironmentVariable("DEFINED_ENVIRONMENT_VARIABLE2", "It's Defined!");
+
+                var tempPath = env.CreateFile(Guid.NewGuid().ToString(), testTargets);
+
+                BuildParameters.WarnOnUninitializedProperty = true;
+
+                MockLogger logger = new MockLogger();
+                logger.Verbosity = LoggerVerbosity.Diagnostic;
+                ProjectCollection pc = new ProjectCollection();
+                pc.RegisterLogger(logger);
+                Project project = pc.LoadProject(tempPath.Path);
+
+                bool result = project.Build();
+                Assert.True(result);
+
+                logger
+                    .AllBuildEvents
+                    .OfType<UninitializedPropertyReadEventArgs>()
+                    .Any(p => p.PropertyName == "DOES_NOT_EXIST")
+                    .ShouldBeTrue();
+
+                logger
+                    .AllBuildEvents
+                    .OfType<EnvironmentVariableReadEventArgs>()
+                    .Any(ev => ev.EnvironmentVariableName == "DEFINED_ENVIRONMENT_VARIABLE2")
+                    .ShouldBeTrue();
+
+                logger
+                    .AllBuildEvents
+                    .OfType<EnvironmentVariableReadEventArgs>()
+                    .Any(ev => ev.EnvironmentVariableName == "DEFINED_ENVIRONMENT_VARIABLE")
```

Comment:
This shouldn't be an environment variable read because it's always overwritten right? So it's not sensitive to whatever value the environment variable had.

Reply:
Ow, SholdBeFalse, nevermind :)

---
In reply to: [297766658](https://github.com/microsoft/msbuild/pull/4461#discussion_r297766658) [](ancestors = 297766658)